service/controller/resource/clusterupgrade/create.go

```diff
@@ -111,18 +111,37 @@ func (r *Resource) ensureMasterHasUpgraded(ctx context.Context, cluster capiv1al
 }
 
 func (r *Resource) propagateReleaseToMachinePools(ctx context.Context, cr capiv1alpha3.Cluster) error {
-	lst := expcapiv1alpha3.MachinePoolList{}
-	err := r.ctrlClient.List(ctx, &lst, client.MatchingLabels{capiv1alpha3.ClusterLabelName: key.ClusterName(&cr)})
+	lst, err := helpers.GetMachinePoolsByMetadata(ctx, r.ctrlClient, cr.ObjectMeta)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	if !hasMachinePoolsUpgradeCompleted(lst.Items, cr.Labels[label.ReleaseVersion]) {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "machinepool still upgrading; skipping release propagation")
-		return nil
-	}
+	for i, mp := range lst.Items {
+		if cr.Labels[label.ReleaseVersion] == mp.Labels[label.ReleaseVersion] &&
+			cr.Labels[label.AzureOperatorVersion] == mp.Labels[label.AzureOperatorVersion] {
+			continue
+		}
+
+		if i > 0 {
+			previousMachinePool := &lst.Items[i-1]
```

Comment:
If possible, I'd propose to avoid anything that is based on sequentiality. Sorting of items is missing here, but even with that the correctness is not guaranteed if `MachinePool` is created with "unlucky" name/ID between reconciliation rounds (newly created CR would become `previousMachinePool`).

Reply:
> Sorting of items is missing here

Yeah, I wasn't sure about this one, thanks for confirming :+1:

Reply:
Here then, in every reconciliation loop, we should check that every node pool is either already upgraded, or the upgraded did not even start. And after checking them all, we take one and update its version labels.

Reply:
Yes, and as I wrote in the other thread, we need to account for situation where `MachinePool` labels are updated to new release, but the conditions are not updated yet. In that case the `Upgrading` should have the previous release version (as discussed last week).

Reply:
> we need to account for situation where MachinePool labels are updated to new release, but the conditions are not updated yet. In that case the Upgrading should have the previous release version (as discussed last week).

Yup, `upgrade.IsNodePoolUpgradeInProgressOrPending(...)` func is checking condition and the actual state.

After updating `MachinePool` labels, the `machinepoolconditions` handler should almost immediately kick in and update the `Upgrading` condition as well, so this time frame where labels are updated, but `Upgrading` condition is still not set to `True`. should really narrow.